File: frontend/types/goals.ts
export interface Goal {
  id: number;
  title: string;
  target_amount: number;
  current_amount: number;
  deadline: string;
  created_at: string;
  updated_at: string;
  remaining: number;
  progress_pct: number;
  months_until_deadline: number;
  monthly_contribution: number;
  months_to_complete: number;
  on_track: boolean;
}

export interface GoalCreateRequest {
  title: string;
  target_amount: number;
  current_amount?: number;
  deadline: string;
}

export interface GoalUpdateRequest {
  title?: string;
  target_amount?: number;
  current_amount?: number;
  deadline?: string;
}

File: backend/app/schemas/goals.py
from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator


def _ensure_future_deadline(value: date) -> date:
    if value < date.today():
        raise ValueError("Deadline must be in the future")
    return value


class GoalCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    target_amount: float = Field(gt=0)
    current_amount: float = Field(default=0.00, ge=0)
    deadline: date

    _validate_deadline = field_validator("deadline")(_ensure_future_deadline)


class GoalUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    target_amount: float | None = Field(default=None, gt=0)
    current_amount: float | None = Field(default=None, ge=0)
    deadline: date | None = None

    _validate_deadline = field_validator("deadline")(_ensure_future_deadline)


class GoalResponse(BaseModel):
    id: int
    title: str
    target_amount: float
    current_amount: float
    deadline: date
    created_at: datetime
    updated_at: datetime
    remaining: float
    progress_pct: float
    months_until_deadline: int
    monthly_contribution: float
    months_to_complete: int
    on_track: bool

    model_config = {"from_attributes": True}
